JavaScript/src/dateTime/baseDuration.ts

```diff
@@ -0,0 +1,272 @@
+    }
+    
+    extract(source: string): Array<ExtractResult> {
+        let baseTokens = this.numberWithUnit(source);
+        let tokens: Array<Token> = new Array<Token>()
+        .concat(baseTokens)
+        .concat(this.numberWithUnitAndSuffix(source, baseTokens))
+        .concat(this.implicitDuration(source))
+        let result = Token.mergeAllTokens(tokens, source, this.extractorName);
+        return result;
+    }
+    
+    private numberWithUnit(source: string): Array<Token> {
+        return this.config.cardinalExtractor.extract(source)
+        .map(o => {
+            let afterString = source.substring(o.start + o.length);
+            let match = RegExpUtility.getMatches(this.config.followedUnit, afterString)[0];
+            if (match && match.index === 0) {
+                return new Token(o.start | 0, o.start + o.length + match.length);
+            }
+        }).filter(o => o !== undefined)
+        .concat(this.getTokensFromRegex(this.config.numberCombinedWithUnit, source))
+        .concat(this.getTokensFromRegex(this.config.anUnitRegex, source))
+        .concat(this.getTokensFromRegex(this.config.inExactNumberUnitRegex, source));
+    }
+    
+    private numberWithUnitAndSuffix(source: string, ers: Token[]): Array<Token> {
+        return ers.map(o => {
+            let afterString = source.substring(o.start + o.length);
+            let match = RegExpUtility.getMatches(this.config.suffixAndRegex, afterString)[0];
+            if (match && match.index === 0) {
+                return new Token(o.start | 0, o.start + o.length + match.length);
+            }
+        });
+    }
+    
+    private implicitDuration(source: string): Array<Token> {
+        return this.getTokensFromRegex(this.config.allRegex, source)
+        .concat(this.getTokensFromRegex(this.config.halfRegex, source));
+    }
+    
+    private getTokensFromRegex(regexp: RegExp, source: string): Array<Token> {
+        return RegExpUtility.getMatches(regexp, source)
+        .map(o => new Token(o.index, o.index + o.length));
+    }
+}
+
+export interface IDurationParserConfiguration {
+    cardinalExtractor: BaseNumberExtractor
+    numberParser: BaseNumberParser
+    followedUnit: RegExp
+    suffixAndRegex: RegExp
+    numberCombinedWithUnit: RegExp
+    anUnitRegex: RegExp
+    allDateUnitRegex: RegExp
+    halfDateUnitRegex: RegExp
+    inExactNumberUnitRegex: RegExp
+    unitMap: ReadonlyMap<string, string>
+    unitValueMap: ReadonlyMap<string, number>
+    doubleNumbers: ReadonlyMap<string, number>
+}
+
+export class BaseDurationParser implements IDateTimeParser {
+    private readonly parserName = Constants.SYS_DATETIME_DURATION;
+    private readonly config: IDurationParserConfiguration;
+    
+    constructor(config: IDurationParserConfiguration) {
+        this.config = config;
+    }
+    
+    parse(extractorResult: ExtractResult, referenceDate?: Date): DateTimeParseResult | null {
+        if (!referenceDate) referenceDate = new Date();
```

Comment:
Nitpick: Add more blank lines throughout the code to improve readability.  Especially between ifs.